pkg/dmaasbackup/veleroschedules.go

```diff
@@ -32,24 +35,90 @@ func (o ScheduleByCreationTimestamp) Less(i, j int) bool {
+// getLatestVeleroSchedule return the latest schedule from status.veleroschedule
+func getLatestVeleroSchedule(dbkp *v1alpha1.DMaaSBackup) *v1alpha1.VeleroScheduleDetails {
 	if len(dbkp.Status.VeleroSchedules) == 0 {
 		return nil
 	}
 
-	// sort veleroschedule with descending creationtimestamp
-	// veleroschedule have only one active schedule
-	sort.Sort(sort.Reverse(ScheduleByCreationTimestamp(dbkp.Status.VeleroSchedules)))
-
 	return &dbkp.Status.VeleroSchedules[0]
 }
 
+// getLastVeleroSchedule return the last created schedule from status.veleroschedule
+func getLastVeleroSchedule(dbkp *v1alpha1.DMaaSBackup) *v1alpha1.VeleroScheduleDetails {
+	if len(dbkp.Status.VeleroSchedules) < 2 {
+		return nil
+	}
+
+	return &dbkp.Status.VeleroSchedules[1]
+}
+
 // appendVeleroSchedule add given schedule to scheduledetails with Active status
+// and sort dbkp.Status.VeleroSchedules using schedule creationTimestamp in reverse order
 func appendVeleroSchedule(dbkp *v1alpha1.DMaaSBackup, schedule *velerov1api.Schedule) {
 	scheduleDetails := v1alpha1.VeleroScheduleDetails{
 		ScheduleName:      schedule.Name,
 		CreationTimestamp: &schedule.CreationTimestamp,
 		Status:            v1alpha1.Active,
 	}
 	dbkp.Status.VeleroSchedules = append(dbkp.Status.VeleroSchedules, scheduleDetails)
+
+	// sort veleroschedule with descending creationtimestamp
+	sort.Sort(sort.Reverse(ScheduleByCreationTimestamp(dbkp.Status.VeleroSchedules)))
+}
+
+// getVeleroScheduleDetails return schedule details from dmaasbackup status for the given schedule
+func getVeleroScheduleDetails(scheduleName string, dbkp *v1alpha1.DMaaSBackup) *v1alpha1.VeleroScheduleDetails {
+	for index, scheduleDetail := range dbkp.Status.VeleroSchedules {
+		if scheduleDetail.ScheduleName == scheduleName {
+			return &dbkp.Status.VeleroSchedules[index]
+		}
+	}
+	return nil
+}
+
+// getDummyVeleroSchedule return the dummy schedule queued for full backup
+func getDummyVeleroSchedule(dbkp *v1alpha1.DMaaSBackup) *v1alpha1.VeleroScheduleDetails {
+	scheduleLen := len(dbkp.Status.VeleroSchedules)
+	if scheduleLen == 0 {
+		return nil
+	}
+
+	lastEntry := &dbkp.Status.VeleroSchedules[scheduleLen-1]
+
+	if lastEntry.CreationTimestamp.IsZero() && lastEntry.Status == "" {
+		return lastEntry
+	}
+
+	return nil
+}
+
+// addDummyVeleroSchedule add the velero schedule with empty timestamp and status
+// and sort dbkp.Status.VeleroSchedules using schedule creationTimestamp in reverse order
+func addDummyVeleroSchedule(dbkp *v1alpha1.DMaaSBackup, name string) {
+	if name == "" {
+		return
+	}
+
+	scheduleDetails := v1alpha1.VeleroScheduleDetails{
+		ScheduleName:      name,
+		CreationTimestamp: &metav1.Time{Time: time.Time{}},
+	}
+
+	dbkp.Status.VeleroSchedules = append(dbkp.Status.VeleroSchedules, scheduleDetails)
+
+	// sort veleroschedule with descending creationtimestamp
+	sort.Sort(sort.Reverse(ScheduleByCreationTimestamp(dbkp.Status.VeleroSchedules)))
```

Comment:
how does Sort works with empty CreationTimestamp? does it make empty one to go to start of the array?

Reply:
we are defining the function for comparison.